lua/dap/ui/variable-tree.lua

```diff
@@ -0,0 +1,246 @@
+      variable_buffers[buf].expanded[v.variablesReference] = true
+      if not v.variables then
+        session:request(
+          "variables",
+          {variablesReference = v.variablesReference},
+          function(_, response)
+            if response then
+              v.variables =
+                utils.map_to_key(
+                function(var)
+                  return var.name
+                end,
+                response.variables
+              )
+              update_variable_buffer(buf, win)
+            end
+          end
+        )
+      end
+    end
+  end
+  update_variable_buffer(buf, win)
+end
+
+-- Variable (https://microsoft.github.io/debug-adapter-protocol/specification#Types_Variable)
+--- name: string
+--- value: string (since we're also accepting scopes here: also optional)
+--- type?: string
+--- presentationHint?: VariablePresentationHint
+--- evaluateName?: string
+--- variablesReference: number
+--- namedVariables?: number
+--- indexedVariables?: number
+--- memoryReference?: string
+function M.write_variables(buf, variables, line, column, win)
+  line = line or 0
+  column = column or 0
+
+  local state = variable_buffers[buf]
+  local win_width = api.nvim_win_get_width(win)
+  local win_height = api.nvim_win_get_height(win)
+  local indent = string.rep(" ", column)
+
+  local sorted_variables = {}
+  for _, v in pairs(variables) do
+    table.insert(sorted_variables, v)
+  end
+  table.sort(
+    sorted_variables,
+    function(a, b)
+      return a.name < b.name
+    end
+  )
+  local max_textlength = 0
+
+  for _, v in pairs(sorted_variables) do
+    state.line_to_variable[line] = v
+    local text =
+      indent ..
+      v.name ..
+        (non_empty_sequence(v.value) and M.variable_value_separator .. v.value or "") ..
+          " " .. (v.type and (" " .. v.type) or "")
+
+    local splitted_text = vim.split(text, "\n")
+    api.nvim_buf_set_lines(buf, line, line + #splitted_text, false, splitted_text)
+
+    if M.multiline_variable_display then
+      line = line + #splitted_text
+    else
+      api.nvim_buf_set_lines(buf, line, line + 1, false, {table.concat(splitted_text, " ")})
+      line = line + 1
+    end
+
+    if v.variables and variable_buffers[buf].expanded[v.variablesReference] then
+      local inner_max
+      line, inner_max = M.write_variables(buf, v.variables, line, column + 2, win)
+      max_textlength = math.max(max_textlength, inner_max)
+    end
+    max_textlength = math.max(max_textlength, #text)
+  end
+  if win then
+    if win_width < max_textlength then
+      api.nvim_win_set_width(win, math.min(max_textlength + 2, M.max_win_width))
+    end
+    if win_height < line then
+      api.nvim_win_set_height(win, line)
+    end
+  end
+  api.nvim_buf_set_lines(buf, line, -1, false, {})
+  return line, max_textlength
+end
+
+function M.hover(session)
+  if not session then
+    return
+  end
+  if not session.stopped_thread_id then
+    return
+  end
```

Comment:
Given that `M.hover` is likely called explicit with a mapping by the user, I think here it would make sense to give some feedback if there is no session or no stopped thread

Reply:
Added a message.